torch/_library/define.py

```diff
@@ -0,0 +1,334 @@
+import inspect
+import sys
+from typing import Callable, Dict, List, Optional, Sequence, Union
+
+from .. import _C, _library, library, Tensor
+
+
+device_types_t = Optional[Union[str, Sequence[str]]]
+
+
+def opaque_op(*, mutated_args: Sequence[str], types: device_types_t = None):
+    """Wraps a function into an opaque custom operator.
+
+    Treats the function as a black-box (that is, PyTorch will never peek into
+    the function). The function must have type hints; these are needed
+    to interface with PyTorch's various subsystems.
+
+    This API is used as a decorator (please see examples).
+
+    Args:
+        mutated_args (Sequence[str]): The names of args that the function mutates.
+            This MUST be accurate, otherwise, the behavior is undefined.
+        types (None | str | Sequence[str]): The device type(s) the function
+            is valid for. If no device type is provided, then the function
+            is used as the default implementation for all device types.
+
+    Examples::
+        >>> import torch
+        >>> from torch import Tensor
+        >>> from torch.library import opaque_op
+        >>> import numpy as np
+        >>>
+        >>> @opaque_op(mutated_args=())
+        >>> def numpy_sin(x: Tensor) -> Tensor:
+        >>>     x_np = x.cpu().numpy()
+        >>>     y_np = np.sin(x_np)
+        >>>     return torch.from_numpy(y_np).to(device=x.device)
+        >>>
+        >>> x = torch.randn(3)
+        >>> y = numpy_sin(x)
+        >>> assert torch.allclose(y, x.sin())
+        >>>
+        >>> # Example of a opaque op that only works for one device type.
+        >>> @opaque_op(mutated_args=(), types="cpu")
+        >>> def numpy_sin_cpu(x: Tensor) -> Tensor:
+        >>>     x_np = x.numpy()
+        >>>     y_np = np.sin(x_np)
+        >>>     return torch.from_numpy(y_np)
+        >>>
+        >>> x = torch.randn(3)
+        >>> y = numpy_sin_cpu(x)
+        >>> assert torch.allclose(y, x.sin())
+
+    """
+    assert len(mutated_args) == 0, "NYI"
+    mod = get_caller_module(stacklevel=2)
+    if mod is None:
+        namespace = "module_not_found"  # can happen in colab
+    else:
+        namespace = mangle_module(mod.__name__)
+
+    def inner(fn):
+        return OpaqueOpDef._from_fn(fn, types, mutated_args, namespace)
+
+    return inner
+
+
+class OpaqueOpDef:
+    def __init__(self, namespace: str, name: str, schema: str):
```

Comment:
Consider adding return type annots here and every where else

Reply:
done